src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -633,60 +632,58 @@ void ReferenceProcessor::balance_queues(DiscoveredList ref_lists[])
   size_t avg_refs = total_refs / _num_queues + 1;
   uint to_idx = 0;
   for (uint from_idx = 0; from_idx < _max_num_queues; from_idx++) {
-    bool move_all = false;
-    if (from_idx >= _num_queues) {
-      move_all = ref_lists[from_idx].length() > 0;
+    size_t from_len = ref_lists[from_idx].length();
+    if (from_len == 0) {
+      // empty
+      continue;
     }
-    while ((ref_lists[from_idx].length() > avg_refs) ||
-           move_all) {
+
+    bool move_all = (from_idx >= _num_queues);
+    size_t remaining_to_move = move_all
+      ? from_len
+      : (from_len > avg_refs ? from_len - avg_refs : 0);
```

Comment:
`move_all` is (now) only used in the immediately following complicated
conditional expression. I think this would be more readable as something like
```
  // If from_idx >= _num_queues, move all.
  // Otherwise, only move excess above avg_refs.
  size_t keep = 0;
  if (from_idx < _num_queues) {
    keep = MIN2(from_len, avg_refs);
  }
  size_t remaining_to_move = from_len - keep;
```

Reply:
I don't think so.

Your suggested comment:

```
// If from_idx >= _num_queues, move all.
// Otherwise, only move excess above avg_refs.
```
can almost directly map to the actual code:

```
    size_t remaining_to_move = move_all
      ? from_len
      : (from_len > avg_refs ? from_len - avg_refs : 0);
```

OTOH, the suggested snippet of using `keep` requires some thinking to connect the dots, IMO.

Reply:
I find nested ternary conditional operators hard to parse, and I seem to not
be alone in that. Some style guides recommend against ternary conditional
operators at all, though I'm not of that camp. Others allow them but call out
nesting them as one kind of problematically complex form, which I agree with.

Perhaps this?

```
size_t remaining_to_move = move_all
    ? from_len
    : (from_len - MIN2(from_len, avg_refs));
```

Reply:
> I find nested ternary conditional operators hard to parse

OK; used `if` to remove the outer layer.